### PR TITLE
Polylines on 3D Tiles

### PR DESCRIPTION
**Pull request by lilleyse**
Follows up on https://github.com/AnalyticalGraphicsInc/cesium/pull/7434 and adds support for polylines on 3D Tiles.

polylines

Demo I've been using for testing: [localhost link](http://localhost:8080/Apps/Sandcastle/index.html#c=7Rhpb9s49q9osx8so46s+3DSYJM0aINpDiTuzodJUdASbWsiiQJFJfEs8t/3kRR1OM40xXTn0woJLPI9vvuiplPtuNDwE8rLDGtkqdVVWqw0pC2cJNdYmlWYaYxocYaqKl1uNFQQtsa0hWNAMO6KB0S1hxQ/AuS9VuBH7RRXaZ0b/xZ7+igWy1NSMJQWmI4m2n/uCg0ehimFrWtKHtIE05k6GFOMGP6V0CyZSxR9fFc8jw8kryrGBQZWkqchlgC6K6agkFYQmqNsICPXrVwTRlYU5TlmdCMJKfBAavnjfJhLoN7IWtOsle+cFDe4IjWNsbGkJD+uAPE80X0zcFtBe8IZJU3zlKUPuDJQkugN3x7aH4Tkc9IHSAkb08eIpaRoJPpCM5B4ZBhT+LsVzvuAGJoOBJ+eypOpPDq9Hqg/Vc77vSLF6KCx3MBkdYWTgfPB81tGVLgDUdOBrG837SuaTiTKFvFNiVtnnL4AGadnt+dfLr45H77Nzz+fNQ7ZKZ9RsU2GX5fyloOVnDHJCATpiK4WSLc9b6KZ8s/wxqM3uH2nBGMVuXOwb4VyvMMNS7Jt3om2qJlGcQFpAwgy5rONxCR5iWgKnm1ysyDF6f/SOdWaPM60JYJ60VjhNY5/q7lfE0Ja/BYVSYwqlmGOPCerVYZPasZIoY9uQaEtg0PRYrTGE21ZFzHf0OM1ju9xMlbF7JX44qTeaw3ygUR93TwS+x8d+vNuaUm2QFSJ29RILcElW0NRrdioJ2cr4MBWq4wssCFOzOHA8QoIVExRAhHejn2g3FVUBOTLyGpLJPCh9u4H2I87vVVruSAJ5lWvUWWN09WazTQTEAVeE0v3BYnvSc0MRlF8r7cnO2JMmg5IJSSuc1wwY4XZWYb568kGivioQRnxQ9t0UVlmm5O0SKBRVh39iaI73ikNcLhaVJg+oAWEd+/USCoyGhtVvahimi6w3jpOwsZDnUHwyzpfQFttwI3106Wmp9UlulT77Tn+UMxqqhz1zEUUAcGrNqKiopfrNAbaKid7u6LD8Q0AoUK1KGNBamGF2xImAmyAV6GbK2lEl67pEsV4SFTQcATJG5TAe6X3JYDgKVYpqxNItOE+pIncNg2zz4Usl7KS/VQmQ9uKuKFAJhPJupMZ+I/HHNOlQBOl/mRQ6lp8fayIK3vmPCIuEKPpU8dArl2hyryTQO9J00sWGeHgiJRtuvFIrNOmJjYxUZJsswJNZv0gWaeYIhqvN7DdE/pa4n5SUP1PLX0MGbzRf9u3DMeKXMeP7CBwLNe0be46PwpD07c9x4xM/jvRJKITeV4Umrbp2lbQIkaWF4RBYNme2yE6lm2bkRc5vkIMgI1ju5EZ+i1B1/FC23ZN13PNsMULrMj3PM+PovDreDzpVM9h3KQpTI0zMVelOVrxEUoo+u0zWZFvp7wZGb+Xq1GbRL0SRUruigqM/ls72j4xTu1UjU8nMDqPGpaAijMcc4SXZZo/0odG4ybj5fTTi8GX48/J1fzTQUeMU8lg8LYN9fYXCHK9e/P7lpJNAX+znn9dtPnZzc3x+eXBz7JdR+8t+jofNNG6/ybHDufan+rinaS58l8PXpk+LnBR603gd5lQkiptc6GZdnYWQMu2PNdxDEhISPLItSB13cDxLNMxIFd9y7dhC7LXsUzPsDzHDewwdFXWvkrUdwDd9D3LiSLPDiVV34wcw3VsJ4ArWiDIWl7oC0awZdv+9+m6NrQe0wyiKHKchiw8huvbkQ2lS5I1QyDrAlLgBGYAl8GvL01jv8E2PtQ0wwtdH2qYFUZKDS8wItuH0giFVarhm6ERchs6VmC+RQ3f45YIQt/1nEDpAUQcKMp+6HNeXI/I9Y3QctwoCC2gu1MR5y2KeLbNGXpQkIGjUsSMDD/0LNeNeNEWioB9wcW2bwVO+H09PB/0BodC3bcCuyHrBZYBAWBCkW/18CIDdkIHYI7T6gEXrusmVbSPGXm8K5rBlPK55nrrJtFJ8VHCm6MtnmqsK0zE1fgcpllUxLga9tKPW2C9VxrU0a0TA37qfP+czH+VdbPufTLEeUwTuB3MNAvCuIM8j5tUVxaHCRcjyoV7OQZwES6aVnnc4vWF6TXSboQRO3KGgVKjbwNa5cANHKGViLdYNdmoDza5+Mhkwq4qsFpdJkBqWNN0jjmBubeqs26EFsffvdtxYRGQ/iwtOR1BCJk7Juk3N6Rn+YPhajwga/8o2dfqc0v9R4ht9fMur5su8v3hUeTNbNjKd4Sg3YtBaEh5OScyoAFJXKc7sApPb/dYtiMSr2rWD8hrSkpM2YvcEJ8RuniUg9zVzfHlx7OtDCGS4K+NJPZu8Okueiefj09/2ZFVgylRGc75afZ1/m9fYd+9yd6h+LR0JPf/leYloYx/vNJhnmc4L+G2BCP9oo7v4bIVV5VK9n+qLwI9tRYovpeNYKaJr08ub7bNv2mE4/7oBZ6Da/BMc8un3vaC0ATTfQq3orrqA5+32KZFWbM+8wcwMiRrto+ydFXMtDxNkgy/5LjPSAlGHHBVoAVhjOR9KLA9nCoTHSbpg5Ym7+/2tj7N3+3Jz1gAWdZZdpv+ge/2jg6ngD84lhHE+VyBrBnacJS1dfRZbhqGcTiF5ctTjcqALmXikKNP4qqtkMW2NAmDSgWHoMWsQAgwQwGrfcvk7QuW6AmWalUxXPIlvEIvQPuLtOAMH1BW41lzm59oYvlFdAsY4gWXUSfMgCuf9DldMACsvD+jKygo+f8L)

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @lilleyse!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by pjcozzi** [comment]
🎉 

In addition to the CZML example, please add/update a Sandcastle example using the CesiumJS API.

**Comment by bagnell** [comment]
This might be an issue that we don't fix. It'll possibly also be an issue with #7434.
image

File: Specs/DataSources/PolylineVisualizerSpec.js [hunk 1]
```diff
@@ -415,6 +417,58 @@ defineSuite([
...
+    function createAndRemoveGeometryWithClassificationType(classificationType) {
+        if (!Entity.supportsPolylinesOnTerrain(scene)) {
+            return;
+        }
+
+        var objects = new EntityCollection();
+        var visualizer = new PolylineVisualizer(scene, objects);
+
+        var polyline = new PolylineGraphics();
+        polyline.positions = new ConstantProperty([Cartesian3.fromDegrees(0.0, 0.0), Cartesian3.fromDegrees(0.0, 1.0)]);
```
[the remark this answers is not in the document]
Comment (lilleyse) [reply Specs/DataSources/PolylineVisualizerSpec.js]: I saw an improvement when changing 1.0 to 0.000001. Master is ~3.5 seconds, this branch is ~5.0, with that change I see ~4.5. It varies a lot though.

[the remark this answers is not in the document]
Comment (OmarShehata) [reply Specs/DataSources/PolylineVisualizerSpec.js]: I haven't confirmed this but I have reason to believe that a spec that takes ~5 seconds on our machines will time out from time to time (like here https://github.com/AnalyticalGraphicsInc/cesium/issues/7249). If there's no easy way to bring this down I'd mention the spec in that issue when that gets merged so I can take a look next time I'm cutting down test times.

File: Source/Scene/GroundPolylinePrimitive.js [hunk 2]
```diff
@@ -483,37 +520,47 @@ define([
...
+            derivedMorphCommand.renderState = groundPolylinePrimitive._renderStateMorph;
+            derivedMorphCommand.shaderProgram = groundPolylinePrimitive._spMorph;
+            derivedMorphCommand.pickId = 'czm_batchTable_pickColor(v_batchId)';
+            command.derivedCommands.colorMorph = derivedMorphCommand;
         }
     }
 
+    function updateAndQueueCommand(groundPolylinePrimitive, command, frameState, modelMatrix, cull, boundingVolume, debugShowBoundingVolume) {
+        // Use derived appearance command for morph and 2D
+        if (frameState.mode === SceneMode.MORPHING && command.shaderProgram !== groundPolylinePrimitive._spMorph) {
```
Comment (bagnell) [inline Source/Scene/GroundPolylinePrimitive.js]: Why does this check the shader? Shouldn't the command always be the original or the tileset command?

File: Source/Scene/GroundPolylinePrimitive.js [hunk 3]
```diff
@@ -530,25 +577,25 @@ define([
...
-                } else if (frameState.mode !== SceneMode.SCENE3D && colorCommand.shaderProgram !== groundPolylinePrimitive._sp2D) {
-                    colorCommand = colorCommand.derivedCommands.color2D;
+                var boundingVolume = boundingSpheres[j];
+                if (queueTerrainCommands) {
+                    command = colorCommands[j];
+                    updateAndQueueCommand(groundPolylinePrimitive, command, frameState, modelMatrix, cull, boundingVolume, debugShowBoundingVolume);
+                }
+                if (queue3DTilesCommands) {
+                    command = colorCommands[j].derivedCommands.tileset;
+                    updateAndQueueCommand(groundPolylinePrimitive, command, frameState, modelMatrix, cull, boundingVolume, debugShowBoundingVolume);
```
Comment (bagnell) [inline Source/Scene/GroundPolylinePrimitive.js]: Wouldn't this throw in 2D? The tileset command doesn't have any derived commands.

File: Specs/DataSources/PolylineVisualizerSpec.js [hunk 4]
```diff
@@ -759,4 +813,110 @@ defineSuite([
...
+                classificationType : ClassificationType.TERRAIN,
+                clampToGround : true
+            }
+        });
+
+        return pollToPromise(function() {
+            scene.initializeFrame();
+            var isUpdated = visualizer.update(time);
+            scene.render(time);
+            return isUpdated;
```
[the remark this answers is not in the document]
Comment (lilleyse) [reply Specs/DataSources/PolylineVisualizerSpec.js]: Yeah, I cleaned it up in `PolylineVisualizerSpec` and `GeometryVisualizerSpec`.

File: Specs/DataSources/PolylineVisualizerSpec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (lilleyse) [reply Specs/DataSources/PolylineVisualizerSpec.js]: Oh sorry for not stating that my times were for the whole file not a single test.

File: Source/Scene/GroundPolylinePrimitive.js (hunk 2 above)
Comment (lilleyse) [reply Source/Scene/GroundPolylinePrimitive.js]: I actually don't know, I copied this from https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/GroundPolylinePrimitive.js#L541-L544. It looks like `GroundPrimitive` doesn't do this sort of check either.

I'll make the change and test that everything still works.

File: Source/Scene/GroundPolylinePrimitive.js (hunk 3 above)
Comment (lilleyse) [reply Source/Scene/GroundPolylinePrimitive.js]: The 2D derived command is created here: https://github.com/AnalyticalGraphicsInc/cesium/pull/7437/files#diff-d3380c9fefd2c0cccbc42b72efa86ce8R536

File: Source/Scene/GroundPolylinePrimitive.js (hunk 2 above)
Comment (lilleyse) [reply Source/Scene/GroundPolylinePrimitive.js]: 2D and morphing seem to work fine still. I pushed the change.

**Comment by lilleyse** [comment]
> In addition to the CZML example, please add/update a Sandcastle example using the CesiumJS API.

I updated the Classification Types demo (the yellow line on the road).

classification

**Comment by lilleyse** [comment]
All the comments should be addressed now. I opened an issue for the slicing through problem: https://github.com/AnalyticalGraphicsInc/cesium/issues/7453. It will require a bit of thinking to get right so I think it should get pushed to some other PR.

**Pull request merged**